torch/lib/THCUNN/generic/Im2Col.cu

```diff
@@ -0,0 +1,201 @@
+#ifndef THC_GENERIC_FILE
+#define THC_GENERIC_FILE "generic/Im2Col.cu"
+#else
+
+static inline void THNN_(Im2Col_shapeCheck)(
+                         THCState *state,
+                         THCTensor *input,
+                         THCTensor *gradOutput,
+                         int kH, int kW, int dH, int dW,
+                         int padH, int padW, int sH, int sW) {
+
+  THArgCheck(kW > 0 && kH > 0, 9,
+             "kernel size should be greater than zero, but got kH: %d kW: %d", kH, kW);
+  THArgCheck(sW > 0 && sH > 0, 11,
+             "stride should be greater than zero, but got sH: %d sW: %d", sH, sW);
+  THArgCheck(dW > 0 && dH > 0, 11,
+             "dilation should be greater than zero, but got dH: %d dW: %d", dH, dW);
+
+  int ndim = input->nDimension;
+  int dimf = 0;
+  int dimh = 1;
+  int dimw = 2;
+
+  if (ndim == 4) {
+      dimf++;
+      dimh++;
+      dimw++;
+  }
+
+  THCUNN_argCheck(state, ndim == 3 || ndim == 4, 2, input,
+                  "3D or 4D input tensor expected but got: %s");
+
+  long nInputPlane  = input->size[dimf];
+  long inputHeight  = input->size[dimh];
+  long inputWidth   = input->size[dimw];
+
+  long outputHeight = (inputHeight + 2*padH - (dH * (kH - 1)) + 1) / sH + 1;
+  long outputWidth  = (inputWidth + 2*padW - (dW * (kW - 1)) + 1) / sW + 1;
+  long nOutputPlane = nInputPlane * kW * kH;
+  long outputLength = outputHeight * outputWidth;
+
+  if (outputWidth < 1 || outputHeight < 1)
+      THError("Given input size: (%d x %d x %d). "
+              "Calculated output size: (%d x %d). Output size is too small",
+              nInputPlane,inputHeight,inputWidth,nOutputPlane,outputLength);
+
+  THCUNN_check_dim_size(state, input, ndim, dimf, nInputPlane);
+
+  if (gradOutput != NULL) {
+    THCUNN_check_dim_size(state, gradOutput, 3, dimf, nOutputPlane);
+    THCUNN_check_dim_size(state, gradOutput, 3, dimh, outputLength);
+  }
+}
+
+void THNN_(Im2Col_updateOutput)(
+           THCState *state,
+           THCTensor *input,
+           THCTensor *output,
+           int kH, int kW,
+           int dH, int dW,
+           int padH, int padW,
+           int sH, int sW) {
+
+  THCUNN_assertSameGPU(state, 2, input, output);
+
+  THNN_(Im2Col_shapeCheck)
+       (state, input, NULL, kH, kW, dH, dW, padH, padW, sH, sW);
+
+  input = THCTensor_(newContiguous)(state, input);
+  int batch = 1;
+  if (input->nDimension == 3) {
+      // Force batch
+      batch = 0;
+      THCTensor_(resize4d)(state, input, 1, input->size[0], input->size[1], input->size[2]);
+  }
+
+  // Batch_size + input planes
```

Comment:
Some of these comments aren't really necessary because it's clear from your code what is happening